soco/core.py

```diff
@@ -1217,29 +1215,43 @@ def switch_to_line_in(self, source=None):
     @property
     def is_playing_radio(self):
         """bool: Is the speaker playing radio?"""
-        response = self.avTransport.GetPositionInfo(
-            [("InstanceID", 0), ("Channel", "Master")]
-        )
-        track_uri = response["TrackURI"]
-        return re.match(r"^x-rincon-mp3radio:", track_uri) is not None
+        return self.music_source == "RADIO"
 
     @property
     def is_playing_line_in(self):
         """bool: Is the speaker playing line-in?"""
-        response = self.avTransport.GetPositionInfo(
-            [("InstanceID", 0), ("Channel", "Master")]
-        )
-        track_uri = response["TrackURI"]
-        return re.match(r"^x-rincon-stream:", track_uri) is not None
+        return self.music_source == "LINE_IN"
 
     @property
     def is_playing_tv(self):
         """bool: Is the playbar speaker input from TV?"""
+        return self.music_source == "TV"
+
+    @property
+    def music_source(self):
```

Comment:
It would be useful to have a `uri` parameter that would override the `GetPositionInfo()` call. This would save a roundtrip when the information is already available, like when receiving an event.

That would also make it possible to decide the type of a favorite URI since one might want to handle each type differently.

Reply:
Well, if this one could be reworked it would not have to be additional.

I guess it comes down to the unanswered question from DPH above, how will it be used. I previously mentioned a case where the current implementation is not helpful: https://github.com/SoCo/SoCo/pull/532#issuecomment-370228251

Reply:
OK, I've added a `music_source_from_uri()` method, and the `music_source` property implementation calls through to this.

Reply:
Awesome!